File: elixir/perfect-numbers/lib/perfect_numbers.ex
defmodule PerfectNumbers do
  @doc """
  Determine the aliquot sum of the given `number`, by summing all the factors
  of `number`, aside from `number` itself.

  Based on this sum, classify the number as:

  :perfect if the aliquot sum is equal to `number`
  :abundant if the aliquot sum is greater than `number`
  :deficient if the aliquot sum is less than `number`
  """
  @spec classify(number :: integer) :: {:ok, atom} | {:error, String.t()}
  def classify(number) when number < 1,
    do: {:error, "Classification is only possible for natural numbers."}

  def classify(number) when number == 1, do: {:ok, :deficient}

  def classify(number) do
    factors = Enum.filter(Enum.to_list(1..trunc(number / 2)), fn x -> rem(number, x) == 0 end)
    aliquot_sum = Enum.sum(factors)

    case number do
      number when number == aliquot_sum -> {:ok, :perfect}
      number when number < aliquot_sum -> {:ok, :abundant}
      number when number > aliquot_sum -> {:ok, :deficient}
    end
  end
end

File: solutions/ruby/perfect-numbers/6/perfect_numbers.rb
class PerfectNumber
    def self.classify(n)
        raise RuntimeError if n <= 0

        case factors_of(n).sum <=> n
        when 0 then "perfect"
        when -1 then "deficient"
        when 1 then "abundant"
        end
    end

    private
    def self.factors_of(n) = (1..n / 2).filter { | i | n % i == 0 }
end